source/common/upstream/cluster_manager_impl.cc

```diff
@@ -104,17 +95,20 @@ void ClusterManagerInitHelper::removeCluster(ClusterManagerCluster& cluster) {
 
   // There is a remote edge case where we can remove a cluster via CDS that has not yet been
   // initialized. When called via the remove cluster API this code catches that case.
-  std::list<ClusterManagerCluster*>* cluster_list;
+  absl::flat_hash_map<std::string, ClusterManagerCluster*>* cluster_map;
   if (cluster.cluster().initializePhase() == Cluster::InitializePhase::Primary) {
-    cluster_list = &primary_init_clusters_;
+    cluster_map = &primary_init_clusters_;
   } else {
     ASSERT(cluster.cluster().initializePhase() == Cluster::InitializePhase::Secondary);
-    cluster_list = &secondary_init_clusters_;
+    cluster_map = &secondary_init_clusters_;
   }
 
   // It is possible that the cluster we are removing has already been initialized, and is not
-  // present in the initializer list. If so, this is fine.
-  cluster_list->remove(&cluster);
+  // present in the initializer map. If so, this is fine.
+  absl::string_view cluster_name = cluster.cluster().info()->name();
+  if (cluster_map->at(cluster_name) == &cluster) {
```

Comment:
How come we now make this check? Was this not necessary previously?

Reply:
This may need to be an ASSERT

Reply:
This code is equivalent to the original: the `remove()` method of `std::list` checks if the argument is identical to any element and removes all of them for which the check is true.

As the comment above suggests it's possible that the cluster we are removing has already been initialized and is not present in the map. Without the check we'd remove a wrong cluster: the one updated once again after it had been initialized. There is a test case that tests this:  `bazel test //test/common/upstream:cluster_manager_impl_test --gtest_filter=*UpdateAlreadyInitialized*`. It starts to fail without the check.

Reply:
Makes sense, thanks for the explanation!

Reply:
Looks like this can lead to a crash when `cluster_name` isn't in the map - see the crash log [here](https://github.com/istio/istio/issues/30588#issuecomment-791059925)

Reply:
@tbarrella if this is crashing can you do a revert PR and we can reapply later?

Reply:
I thought the comment was about the case when a cluster gets removed, but another one with the same name still exists.

Anyway I should have checked the code was exception safe. Will submit a new fix shortly with a test checking for removal of an unknown cluster.